File: people/thom/test/opsmain.f
PROGRAM opsmain 
	REAL, DIMENSION (0:23) :: in1 = (/ 0,1,2,3,4,5,6,7,8,9,10,11,0,1,2,3,4,5,6,7,8,9,10,11 /)
	REAL, DIMENSION (0:23) :: in2 = (/ 11,10,9,8,7,6,5,4,3,2,1,0,11,10,9,8,7,6,5,4,3,2,1,0 /)
	REAL, DIMENSION (0:23) :: in3 = (/ 5,9,23,-3,5,3,-22,3,5,9,10,-3,9,2,3,5,-2,3,-3,-5,1,1,1,1 /)
	REAL, DIMENSION (0:23) :: out
	INTERFACE
		SUBROUTINE do_ops (mdat, cdat, in1, in2, in3, out)
			CHARACTER(LEN=*) :: mdat, cdat
			REAL, DIMENSION (0:) :: in1, in2, in3, out
		END SUBROUTINE do_ops
	END INTERFACE

	CALL do_ops ("opsm.dat", "opsf.dat", in1, in2, in3, out)
END PROGRAM opsmain

SUBROUTINE do_ops (mdat, cdat, in1, in2, in3, out)
	CHARACTER(LEN=*) :: mdat, cdat
	REAL, DIMENSION (0:) :: in1, in2, in3, out
	INTEGER :: i, j, k

	INTERFACE 
		SUBROUTINE ADD_SUB1(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB1
	END INTERFACE

	INTERFACE 
		SUBROUTINE ADD_SUB2(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB2
	END INTERFACE

	INTERFACE 
		SUBROUTINE ADD_SUB3(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB3
	END INTERFACE

	INTERFACE 
		SUBROUTINE ADD_SUB1_F90(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB1_F90
	END INTERFACE

	INTERFACE 
		SUBROUTINE ADD_SUB2_F90(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB2_F90
	END INTERFACE

	INTERFACE 
		SUBROUTINE ADD_SUB3_F90(f90_in1, f90_in2, f90_in3, f90_out)
			REAL, DIMENSION (0:) :: f90_in1
			REAL, DIMENSION (0:) :: f90_in2
			REAL, DIMENSION (0:) :: f90_in3
			REAL, DIMENSION (0:) :: f90_out
		END SUBROUTINE ADD_SUB3_F90
	END INTERFACE


	OPEN (1, FILE=mdat, STATUS="REPLACE", ACCESS="SEQUENTIAL", ACTION="WRITE")

	PRINT *, "HERE"
	CALL ADD_SUB1 (in1, in2, in3, out)
	PRINT *, "HERE"
	WRITE (1,"(6(4F8.3, /))") out
	WRITE (1,"(A)") ""

	PRINT *, "HERE"
	CALL ADD_SUB2 (in1, in2, in3, out)
	PRINT *, "HERE"
	WRITE (1,"(6(4F8.3, /))") out
	WRITE (1,"(A)") ""

	PRINT *, "HERE"
	CALL ADD_SUB3 (in1, in2, in3, out)
	PRINT *, "HERE"
	DO i=0, 1
		DO j=0, 2
			WRITE (1,"(4F8.3)") out(i*12+j*4:i*12+j*4+3)
		END DO
	END DO

	CLOSE (1)

	OPEN (1, FILE=cdat, STATUS="REPLACE", ACCESS="SEQUENTIAL", ACTION="WRITE")
!	if ((fout = fopen (mdat, "w"))==NULL)
!	{
!		printf("Cannot opsn %s.\n",mdat)
!		exit(0)
!	}

	PRINT *, "HERE"
	CALL ADD_SUB1_F90 (in1, in2, in3, out)
	PRINT *, "HERE"
	DO i=0, 1
		DO j=0, 2
			WRITE (1,"(4F8.3)") out(i*12+j*4:i*12+j*4+3)
		END DO
	END DO
	WRITE (1,"(A)") ""

	PRINT *, "HERE"
	CALL ADD_SUB2_F90 (in1, in2, in3, out)
	PRINT *, "HERE"
	DO i=0, 1
		DO j=0, 2
			WRITE (1,"(4F8.3)") out(i*12+j*4:i*12+j*4+3)
		END DO
	END DO
	WRITE (1,"(A)") ""

	PRINT *, "HERE"
	CALL ADD_SUB3_F90 (in1, in2, in3, out)
	PRINT *, "HERE"
	DO i=0, 1
		DO j=0, 2
			WRITE (1,"(4F8.3)") out(i*12+j*4:i*12+j*4+3)
		END DO
	END DO

	CLOSE (1)

END SUBROUTINE do_ops

File: people/thom/test/bak/opsmain.c
#include <stdio.h>

do_ops (char *mdat, char *cdat, double *in1, double *in2, double *in3,
        double *out)
{
	FILE	*fout, *gout;
	int 	i, j, k;

	if ((fout = fopen (mdat, "w"))==NULL)
	{
		printf("Cannot open %s\n",mdat);
		exit(0);
	}

	ADD_SUB1 (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fprintf (fout, "\n\n");

	ADD_SUB2 (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fprintf (fout, "\n\n");

	ADD_SUB3 (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fclose (fout);

	if ((fout = fopen (cdat, "w"))==NULL)
	{
		printf("Cannot open %s\n",cdat);
		exit(0);
	}

	ADD_SUB1_C (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fprintf (fout, "\n\n");

	ADD_SUB2_C (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fprintf (fout, "\n\n");

	ADD_SUB3_C (in1, in2, in3, out);
	for (i=0; i<2; i++) {
		for (j=0; j<3; j++) {
			for (k=0; k<4; k++) {
				fprintf(fout, "%f ",out[i*12+j*4+k]);
			}
			fprintf (fout, "\n");
		}
		fprintf(fout, "\n");
	}
	fclose (fout);
}

main (int argc, char *argv[])
{
	double	in1[24]={0,1,2,3,4,5,6,7,8,9,10,11,0,1,2,3,4,5,6,7,8,9,10,11};
	double	in2[24]={11,10,9,8,7,6,5,4,3,2,1,0,11,10,9,8,7,6,5,4,3,2,1,0};
	double	in3[24]={5,9,23,-3,5,3,-22,3,5,9,10,-3,9,2,3,5,-2,3,-3,-5,1,1,1,1};
	double	out[24];

	if (argc != 3)
	{
		printf("usage: %s <moa output file> <c output file>\n",argv[0]);
		exit(0);
	}

	do_ops (argv[1], argv[2], in1, in2, in3, out);
}
